DATA RACE

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions.-->

**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/.):

**What keywords did you search in NGINX Ingress controller issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as you can.  If you leave out
    information, we can't help you as well.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
-->

**NGINX Ingress controller version**:
0.9-beta.19

**Kubernetes version** (use `kubectl version`):


**Environment**:

local mac machine


**What happened**:
Run `make test`

**What you expected to happen**:

```
=== RUN   TestStatusActions
==================
WARNING: DATA RACE
Write at 0x00c4201a0b80 by goroutine 24:
  k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).tryAcquireOrRenew()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:231 +0x34b
  k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).acquire.func1()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:176 +0x64
  k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x6f
  k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xcd
  k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).acquire()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:175 +0x135
  k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:143 +0x66

Previous read at 0x00c4201a0b80 by goroutine 26:
  k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection.(*LeaderElector).IsLeader()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:168 +0x7b
  k8s.io/ingress-nginx/internal/ingress/status.(*statusSync).sync()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/ingress/status/status.go:149 +0x83
  k8s.io/ingress-nginx/internal/ingress/status.(*statusSync).(k8s.io/ingress-nginx/internal/ingress/status.sync)-fm()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/ingress/status/status.go:179 +0x55
  k8s.io/ingress-nginx/internal/task.(*Queue).worker()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/task/queue.go:112 +0x4ca
  k8s.io/ingress-nginx/internal/task.(*Queue).(k8s.io/ingress-nginx/internal/task.worker)-fm()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/task/queue.go:59 +0x41
  k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x6f
  k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xcd
  k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x5a
  k8s.io/ingress-nginx/internal/task.(*Queue).Run()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/task/queue.go:59 +0x94

Goroutine 24 (running) created at:
  k8s.io/ingress-nginx/internal/ingress/status.statusSync.Run()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/ingress/status/status.go:95 +0xa6

Goroutine 26 (running) created at:
  k8s.io/ingress-nginx/internal/ingress/status.statusSync.Run()
      /Users/xianlu/WorkSpace/golang/src/k8s.io/ingress-nginx/internal/ingress/status/status.go:97 +0x186
==================
E1207 15:37:46.932614   38411 event.go:260] Could not construct reference to: '&v1.ConfigMap{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"-nginx", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string{"control-plane.alpha.kubernetes.io/leader":"{\"holderIdentity\":\"foo1\",\"leaseDurationSeconds\":30,\"acquireTime\":\"2017-12-07T07:37:46Z\",\"renewTime\":\"2017-12-07T07:37:46Z\",\"leaderTransitions\":0}"}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Data:map[string]string(nil)}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Normal' 'LeaderElection' 'foo1 became leader'
--- FAIL: TestStatusActions (0.11s)
	testing.go:699: race detected during execution of test

```

--- a/internal/ingress/controller/nginx.go
+++ b/internal/ingress/controller/nginx.go
@@ -268,7 +268,7 @@ func (n *NGINXController) Start() {
 	}
 
 	if n.syncStatus != nil {
-		go n.syncStatus.Run(n.stopCh)
+		go n.syncStatus.Run()
 	}
 
 	go wait.Until(n.checkMissingSecrets, 30*time.Second, n.stopCh)

--- a/internal/ingress/status/status.go
+++ b/internal/ingress/status/status.go
@@ -52,7 +52,7 @@ const (
 
 // Sync ...
 type Sync interface {
-	Run(stopCh <-chan struct{})
+	Run()
 	Shutdown()
 }
 
@@ -91,16 +91,8 @@ type statusSync struct {
 }
 
 // Run starts the loop to keep the status in sync
-func (s statusSync) Run(stopCh <-chan struct{}) {
-	go s.elector.Run()
-	go wait.Forever(s.update, updateInterval)
-	go s.syncQueue.Run(time.Second, stopCh)
-	<-stopCh
-}
-
-func (s *statusSync) update() {
-	// send a dummy object to the queue to force a sync
-	s.syncQueue.Enqueue("sync status")
+func (s statusSync) Run() {
+	s.elector.Run()
 }
 
 // Shutdown stop the sync. In case the instance is the leader it will remove the current IP
@@ -146,11 +138,6 @@ func (s *statusSync) sync(key interface{}) error {
 		return nil
 	}
 
-	if !s.elector.IsLeader() {
-		glog.V(2).Infof("skipping Ingress status update (I am not the current leader)")
-		return nil
-	}
-
 	addrs, err := s.runningAddresses()
 	if err != nil {
 		return err
@@ -188,6 +175,12 @@ func NewStatusSyncer(config Config) Sync {
 	callbacks := leaderelection.LeaderCallbacks{
 		OnStartedLeading: func(stop <-chan struct{}) {
 			glog.V(2).Infof("I am the new status update leader")
+			go st.syncQueue.Run(time.Second, stop)
+			wait.PollUntil(updateInterval, func() (bool, error) {
+				// send a dummy object to the queue to force a sync
+				st.syncQueue.Enqueue("sync status")
+				return false, nil
+			}, stop)
 		},
 		OnStoppedLeading: func() {
 			glog.V(2).Infof("I am not status update leader anymore")

--- a/internal/ingress/status/status_test.go
+++ b/internal/ingress/status/status_test.go
@@ -273,9 +273,8 @@ func TestStatusActions(t *testing.T) {
 
 	fk := fkSync.(statusSync)
 
-	ns := make(chan struct{})
 	// start it and wait for the election and syn actions
-	go fk.Run(ns)
+	go fk.Run()
 	//  wait for the election
 	time.Sleep(100 * time.Millisecond)
 	// execute sync
@@ -293,6 +292,8 @@ func TestStatusActions(t *testing.T) {
 	if !ingressSliceEqual(fooIngress1CurIPs, newIPs) {
 		t.Fatalf("returned %v but expected %v", fooIngress1CurIPs, newIPs)
 	}
+
+	time.Sleep(1 * time.Second)
 
 	// execute shutdown
 	fk.Shutdown()
@@ -314,9 +315,6 @@ func TestStatusActions(t *testing.T) {
 	if oic.Status.LoadBalancer.Ingress[0].IP != "0.0.0.0" && oic.Status.LoadBalancer.Ingress[0].Hostname != "foo.bar.com" {
 		t.Fatalf("invalid ingress status for rule with different class")
 	}
-
-	// end test
-	ns <- struct{}{}
 }
 
 func TestCallback(t *testing.T) {
@@ -325,6 +323,7 @@ func TestCallback(t *testing.T) {
 
 func TestKeyfunc(t *testing.T) {
 	fk := buildStatusSync()
+
 	i := "foo_base_pod"
 	r, err := fk.keyfunc(i)
 

